containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2837 +275,1076 @@
+#if MIN_VERSION_base(4,6,0)
+    foldr' = foldr'
+    foldl' = foldl'
 #endif
 
--- | Traverses in order of increasing key.
 instance Traversable IntMap where
-    traverse f = traverseWithKey (\_ -> f)
-    {-# INLINE traverse #-}
+    traverse f = start
+      where
+        start (IntMap Empty) = pure (IntMap Empty)
+        start (IntMap (NonEmpty min minV node)) = liftA2 (\minV' root' -> IntMap (NonEmpty min minV' root')) (f minV) (goL node)
 
-instance NFData a => NFData (IntMap a) where
-    rnf Nil = ()
-    rnf (Tip _ v) = rnf v
-    rnf (Bin _ _ l r) = rnf l `seq` rnf r
+        goL Tip = pure Tip
+        goL (Bin max maxV l r) = liftA3 (\l' r' v' -> Bin max v' l' r') (goL l) (goR r) (f maxV)
+
+        goR Tip = pure Tip
+        goR (Bin min minV l r) = liftA3 (Bin min) (f minV) (goL l) (goR r)
 
-#if __GLASGOW_HASKELL__
+instance Monoid (IntMap a) where
+    mempty = empty
+    mappend = union
+
+#if MIN_VERSION_base(4,9,0)
+instance Semigroup (IntMap a) where
+    (<>) = union
+    stimes = stimesIdempotentMonoid
+#endif
 
-{--------------------------------------------------------------------
-  A Data instance
---------------------------------------------------------------------}
+instance NFData a => NFData (IntMap a) where
+    rnf (IntMap Empty) = ()
+    rnf (IntMap (NonEmpty _ v n)) = rnf v `seq` rnf n
 
--- This instance preserves data abstraction at the cost of inefficiency.
--- We provide limited reflection services for the sake of data abstraction.
+instance NFData a => NFData (Node t a) where
+    rnf Tip = ()
+    rnf (Bin _ v l r) = rnf v `seq` rnf l `seq` rnf r
 
-instance Data a => Data (IntMap a) where
-  gfoldl f z im = z fromList `f` (toList im)
-  toConstr _     = fromListConstr
-  gunfold k z c  = case constrIndex c of
-    1 -> k (z fromList)
-    _ -> error "gunfold"
-  dataTypeOf _   = intMapDataType
-  dataCast1 f    = gcast1 f
+-- | /O(min(n,W))/. Find the value at a key.
+-- Calls 'error' when the element can not be found.
+--
+-- > fromList [(5,'a'), (3,'b')] ! 1    Error: element not in the map
+-- > fromList [(5,'a'), (3,'b')] ! 5 == 'a'
+(!) :: IntMap a -> Key -> a
+(!) m k = findWithDefault (error $ "IntMap.!: key " ++ show k ++ " is not an element of the map") k m
 
-fromListConstr :: Constr
-fromListConstr = mkConstr intMapDataType "fromList" [] Prefix
 
-intMapDataType :: DataType
-intMapDataType = mkDataType "Data.IntMap.Internal.IntMap" [fromListConstr]
+-- | /O(min(n,W))/. Find the value at a key.
+-- Returns 'Nothing' when the element can not be found.
+--
+-- > fromList [(5,'a'), (3,'b')] !? 1 == Nothing
+-- > fromList [(5,'a'), (3,'b')] !? 5 == Just 'a'
+--
+-- @since 0.5.11
+(!?) :: IntMap a -> Key -> Maybe a
+(!?) m k = lookup k m
 
-#endif
+-- | Same as 'difference'.
+(\\) :: IntMap a -> IntMap b -> IntMap a
+(\\) = difference
 
-{--------------------------------------------------------------------
-  Query
---------------------------------------------------------------------}
```

Comment:
Wow, I just assumed that the diff was messy and meaningless enough (since it is a rewrite) that no one would actually look at it. I can maybe add these back in? I forget whether I organized functions in the exact same way, and I certainly moved a bunch of functions to `Data.IntMap.Lazy`.

Reply:
It's probably not worth a lot of effort.